packages/@react-aria/menu/src/useSubmenuTrigger.ts

```diff
@@ -0,0 +1,260 @@
+  // to the menu since this is too far up . Maybe just provide the subMenuTriggerState and have useMenu and useMenuItem handle the key handlers internally?
+  // Maybe we should just have the user pass something like onMenuClose to the trigger level since we have the same problem of being unable to call onClose if the user interacts outside?
+  // Maybe have Menu call onClose in an effect when it detects that it is unmounting? StrictMode is problematic since the cleanup fires even when the component isn't actually unmounting
+  // (not a problem if we wanna keep onClose to fire only if the user selects a item)
+  let subMenuKeyDown = (e: KeyboardEvent) => {
+    switch (e.key) {
+      case 'ArrowLeft':
+        if (direction === 'ltr') {
+          onSubmenuClose();
+        }
+        break;
+      case 'ArrowRight':
+        if (direction === 'rtl') {
+          onSubmenuClose();
+        }
+        break;
+      case 'Escape':
+        state.closeAll();
+        break;
+    }
+  };
+
+  let subMenuProps = {
+    id: overlayId,
+    'aria-label': node.textValue,
+    level: state.level,
+    ...(subMenuType === 'menu' && {
+      onClose: state.closeAll,
+      autoFocus: state.focusStrategy,
+      onKeyDown: subMenuKeyDown
+    })
+  };
+
+  let subMenuTriggerKeyDown = (e: KeyboardEvent) => {
+    switch (e.key) {
+      case 'ArrowRight':
+        if (!isDisabled) {
+          if (direction === 'ltr') {
+            if (subMenuType === 'menu' && !!subMenuRef?.current && document.activeElement === ref?.current) {
+              subMenuRef.current.focus();
+            } else {
+              onSubmenuOpen('first');
+            }
+          } else if (state.isOpen) {
+            e.stopPropagation();
+            onSubmenuClose();
+          }
+        }
+
+        break;
+      case 'ArrowLeft':
+        if (!isDisabled) {
+          if (direction === 'rtl') {
+            if (subMenuType === 'menu' && !!subMenuRef?.current && document.activeElement === ref?.current) {
+              subMenuRef.current.focus();
+            } else {
+              onSubmenuOpen('first');
+            }
+          } else if (state.isOpen) {
+            e.stopPropagation();
+            onSubmenuClose();
+          }
+        }
+        break;
+    }
+  };
+
+  let onPressStart = (e: PressEvent) => {
+    if (!isDisabled && (e.pointerType === 'virtual' || e.pointerType === 'keyboard')) {
+      // If opened with a screen reader or keyboard, auto focus the first submenu item.
+      onSubmenuOpen('first');
+    }
+  };
+
+  let onPress = (e: PressEvent) => {
+    if (!isDisabled && e.pointerType === 'touch') {
+      onSubmenuOpen();
+    }
+  };
+
+  let onHoverChange = (isHovered) => {
+    if (!isDisabled) {
+      if (isHovered && !state.isOpen) {
+        if (!openTimeout.current) {
+          openTimeout.current = setTimeout(() => {
+            onSubmenuOpen();
+          }, 200);
```

Comment:
Do we need the timeout anymore with the safe triangle?

Reply:
It still serves as a warmup timer so a submenu doesn't "flash" open when a user moves their pointer through the menu items quickly.